src/HealthGPS/risk_factor_adjustable_model.cpp

```diff
@@ -0,0 +1,170 @@
+    RiskFactorSexAgeTable adjustments;
+
+    // Baseline scenatio: compute adjustments.
+    if (context.scenario().type() == ScenarioType::baseline) {
+        adjustments = calculate_adjustments(context, keys);
+    }
+
+    // Intervention scenario: recieve adjustments from baseline scenario.
+    else {
+        auto message = context.scenario().channel().try_receive(context.sync_timeout_millis());
+        if (!message.has_value()) {
+            throw core::HgpsException(
+                "Simulation out of sync, receive baseline adjustments message has timed out");
+        }
+
+        auto &basePtr = message.value();
+        auto *messagePrt = dynamic_cast<RiskFactorAdjustmentMessage *>(basePtr.get());
+        if (!messagePrt) {
+            throw core::HgpsException(
+                "Simulation out of sync, failed to receive a baseline adjustments message");
+        }
+
+        adjustments = messagePrt->data();
+    }
+
+    // All scenarios: apply adjustments to population.
+    auto &pop = context.population();
+    std::for_each(core::execution_policy, pop.begin(), pop.end(), [&](auto &person) {
+        if (!person.is_active()) {
+            return;
+        }
+
+        for (const auto &factor : keys) {
+            double delta = adjustments.at(person.gender, factor).at(person.age);
+            person.risk_factors.at(factor) += delta;
+        }
+    });
+
+    // Baseline scenario: send adjustments to intervention scenario.
+    if (context.scenario().type() == ScenarioType::baseline) {
+        context.scenario().channel().send(std::make_unique<RiskFactorAdjustmentMessage>(
+            context.current_run(), context.time_now(), std::move(adjustments)));
+    }
+}
+
+RiskFactorSexAgeTable
+RiskFactorAdjustableModel::calculate_adjustments(RuntimeContext &context,
+                                                 const std::vector<core::Identifier> &keys) const {
+    auto age_range = context.age_range();
+    auto max_age = age_range.upper() + 1;
+
+    // Compute simulated means.
+    auto simulated_means = calculate_simulated_mean(context, keys);
+
+    // Compute adjustments.
+    auto adjustments = RiskFactorSexAgeTable{};
+    for (const auto &[sex, simulated_means_by_sex] : simulated_means) {
+        for (const auto &factor : keys) {
+            adjustments.emplace(sex, factor, std::vector<double>(max_age, 0.0));
+            for (auto age = age_range.lower(); age <= age_range.upper(); age++) {
+                double expect = risk_factor_expected_.at(sex, factor).at(age);
+                double sim_mean = simulated_means_by_sex.at(factor).at(age);
+                double delta = expect - sim_mean;
+                if (!std::isnan(delta)) {
+                    adjustments.at(sex, factor).at(age) = delta;
+                }
+            }
+        }
+    }
+
+    return adjustments;
+}
+
+RiskFactorSexAgeTable
+RiskFactorAdjustableModel::calculate_simulated_mean(RuntimeContext &context,
```

Comment:
Out of `context`, this function only needs the age range and the population. Then, why passing the whole of the `context` object around? This probably also applies to other functions. 

Passing `context` hides what this function really depends on, which is not the whole context, but only those two specific values. It is way more explicit to pass that rather than a hold-it-all object whose internal structure might be unknown.

Reply:
I habitually pass things around in packaged form, but I see your point. I'll change the ones I spot.

Reply:
I've only addressed this instance, but I'll keep this in mind going forwards.